modules/platform/nuxeo-thumbnail/src/main/java/org/nuxeo/ecm/platform/thumbnail/listener/ThumbnailHelper.java

```diff
@@ -0,0 +1,152 @@
+ * limitations under the License.
+ *
+ * Contributors:
+ * Vladimir Pasquier <vpasquier@nuxeo.com>
+ * Laurent Doguin <ldoguin@nuxeo.com>
+ * Nelson Silva <nsilva@nuxeo.com>
+ * bdelbosc
+ */
+package org.nuxeo.ecm.platform.thumbnail.listener;
+
+import static org.nuxeo.ecm.core.api.CoreSession.ALLOW_VERSION_WRITE;
+import static org.nuxeo.ecm.platform.thumbnail.listener.UpdateThumbnailListener.THUMBNAIL_UPDATED;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.Serializable;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.nuxeo.ecm.core.api.Blob;
+import org.nuxeo.ecm.core.api.Blobs;
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.NuxeoException;
+import org.nuxeo.ecm.core.api.VersioningOption;
+import org.nuxeo.ecm.core.api.blobholder.BlobHolder;
+import org.nuxeo.ecm.core.api.thumbnail.ThumbnailAdapter;
+import org.nuxeo.ecm.core.api.versioning.VersioningService;
+import org.nuxeo.ecm.core.blob.BlobManager;
+import org.nuxeo.ecm.platform.dublincore.listener.DublinCoreListener;
+import org.nuxeo.ecm.platform.ec.notification.NotificationConstants;
+import org.nuxeo.ecm.platform.thumbnail.ThumbnailConstants;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.transaction.TransactionHelper;
+
+/**
+ * Helper to create document thumbnail if necessary.
+ *
+ * @since 11.5
+ */
+public class ThumbnailHelper {
+
+    private static final Logger log = LogManager.getLogger(ThumbnailHelper.class);
+
+    public static final String THUMBNAIL_TX_TIMEOUT_PROPERTY = "nuxeo.thumbnail.transaction.timeout.seconds";
+
+    public static final int DEFAULT_TX_TIMEOUT_SECONDS = 300;
+
+    protected Integer transactionTimeout;
+
+    /**
+     * Creates a thumbnail if needed and sync the document facet.
+     */
+    public void createThumbnailIfNeeded(CoreSession session, DocumentModel doc) {
+        Blob thumbnailBlob = getManagedThumbnail(doc);
+        if (thumbnailBlob != null) {
+            log.debug("Found a managed thumbnail for doc: {}", doc::getId);
+        } else {
+            ThumbnailAdapter thumbnailAdapter = doc.getAdapter(ThumbnailAdapter.class);
+            if (thumbnailAdapter == null) {
+                log.debug("No thumbnail adapter for doc: {}, skipping", doc::getId);
+                return;
+            }
+            log.debug("Fetching or building a thumbnail for doc: {}, session ttl: {}s", doc::getId,
+                    TransactionHelper::getTransactionTimeToLive);
+            thumbnailBlob = thumbnailAdapter.computeThumbnail(session);
+        }
+        boolean forceUpdate = false;
+        if (thumbnailBlob == null) {
+            log.debug("Failed to create thumbnail for doc: {}", doc::getId);
+            if (doc.hasFacet(ThumbnailConstants.THUMBNAIL_FACET)) {
+                doc.removeFacet(ThumbnailConstants.THUMBNAIL_FACET);
```

Comment:
Here in the original code we were nullifying the thumbnail property, see https://github.com/nuxeo/nuxeo/pull/4839/commits/bb8ecd23205c354b5226c1ec64c5734148a635c3#diff-fc9133ce972bfcea5e4fba9d87a7a6f15f97acbb31bee9a177dadf7132dd2ca3L106.
Don't we want to do the same thing?

Reply:
this is not necessary because the property is part of schema facet that we remove